File: frontend/src/services/album.ts
import api from '@/services/api'

export interface AlbumPhoto {
  id: number
  image_url: string
  thumbnail_url: string
  width: number
  height: number
  caption: string | null
  location: string | null
  taken_on: string | null
  alt_text: string | null
  show_in_carousel: boolean
  show_in_gallery: boolean
  carousel_order: number
  gallery_order: number
  published: boolean
  created_at: string
  updated_at: string
}

export interface AlbumFeed {
  carousel: AlbumPhoto[]
  gallery: AlbumPhoto[]
  autoplay_delay_ms: number
}

export interface AlbumAdminFeed {
  items: AlbumPhoto[]
  autoplay_delay_ms: number
}

export async function fetchAlbum(signal?: AbortSignal): Promise<AlbumFeed> {
  const response = await api.get<AlbumFeed>('/album', { signal })
  return response.data
}

export async function fetchAdminAlbum(signal?: AbortSignal): Promise<AlbumAdminFeed> {
  const response = await api.get<AlbumAdminFeed>('/admin/album', { signal })
  return response.data
}

File: backend/app/schemas/album.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class AlbumPhotoUpdate(BaseModel):
    caption: str | None = Field(default=None, max_length=1000)
    location: str | None = Field(default=None, max_length=200)
    taken_on: date | None = None
    alt_text: str | None = Field(default=None, max_length=300)
    show_in_carousel: bool | None = None
    show_in_gallery: bool | None = None
    carousel_order: int | None = Field(default=None, ge=0, le=1_000_000)
    gallery_order: int | None = Field(default=None, ge=0, le=1_000_000)
    published: bool | None = None


class AlbumPhotoRotate(BaseModel):
    degrees: Literal[-90, 90, 180]


class AlbumSettingsUpdate(BaseModel):
    autoplay_delay_ms: int = Field(ge=2000, le=20000)


class AlbumPhotoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    image_url: str
    thumbnail_url: str
    width: int
    height: int
    caption: str | None = None
    location: str | None = None
    taken_on: date | None = None
    alt_text: str | None = None
    show_in_carousel: bool
    show_in_gallery: bool
    carousel_order: int
    gallery_order: int
    published: bool
    created_at: str
    updated_at: str


class AlbumFeedResponse(BaseModel):
    carousel: list[AlbumPhotoResponse]
    gallery: list[AlbumPhotoResponse]
    autoplay_delay_ms: int


class AlbumAdminResponse(BaseModel):
    items: list[AlbumPhotoResponse]
    autoplay_delay_ms: int
